engine/cdn/item_logs_handler.go

```diff
@@ -75,6 +82,103 @@ func (s *Service) getItemDownloadHandler() service.Handler {
 	}
 }
 
+func (s *Service) getItemLogsStreamHandler() service.Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		vars := mux.Vars(r)
+		itemType := sdk.CDNItemType(vars["type"])
+		if !itemType.IsLog() {
+			return sdk.NewErrorFrom(sdk.ErrWrongRequest, "invalid item log type")
+		}
+		apiRef := vars["apiRef"]
+
+		offset := service.FormInt64(r, "offset")
+
+		it, err := item.LoadByAPIRefHashAndType(ctx, s.Mapper, s.mustDBWithCtx(ctx), apiRef, itemType)
+		if err != nil {
+			return err
+		}
+		iu, err := storage.LoadItemUnitByUnit(ctx, s.Mapper, s.mustDBWithCtx(ctx), s.Units.Buffer.ID(), it.ID)
+		if err != nil {
+			return err
+		}
+
+		c, err := websocket.Upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			service.WriteError(ctx, w, r, sdk.NewErrorWithStack(err, sdk.ErrWebsocketUpgrade))
+			return nil
+		}
+		defer c.Close()
+
+		wsClient := websocket.NewClient(c)
+		wsClientData := &websocketClientData{
+			itemID:              it.ID,
+			chanItemUpdate:      make(chan struct{}, 10),
+			scoreNextLineToSend: offset,
+		}
+		s.WSServer.AddClient(wsClient, wsClientData)
+		defer s.WSServer.RemoveClient(wsClient.UUID())
+
+		s.GoRoutines.Exec(ctx, "getItemLogsStreamHandler."+wsClient.UUID(), func(ctx context.Context) {
+			log.Debug("getItemLogsStreamHandler> start routine for client %s", wsClient.UUID())
+
+			send := func() error {
+				log.Debug("getItemLogsStreamHandler> send log to client %s from %d", wsClient.UUID(), wsClientData.scoreNextLineToSend)
+
+				rc, err := s.Units.Buffer.NewAdvancedReader(ctx, *iu, sdk.CDNReaderFormatJSON, wsClientData.scoreNextLineToSend, 100)
+				if err != nil {
+					return err
+				}
+				defer rc.Close() // nolint
+				buf := new(bytes.Buffer)
+				if _, err := io.Copy(buf, rc); err != nil {
+					return sdk.WrapError(err, "cannot copy data from reader to memory buffer")
+				}
+				var lines []redis.Line
+				if err := json.Unmarshal(buf.Bytes(), &lines); err != nil {
```

Comment:
is it mandatory to marshal the content ? can't read from `rc` and write to the `websocket`?